### PR TITLE
Feat/contour

### PR DESCRIPTION
**Pull request by purpose233**
[[中文版模板 / Chinese template](https://github.com/VisActor/vgrammar/blob/main/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Refactoring
- [ ] Update dependency
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Workflow
- [ ] Chore
- [ ] Release
- [ ] Other (about what?)

### 🔗 Related issue link

[Contour Transform](https://github.com/orgs/VisActor/projects/20/views/5?pane=issue&itemId=32374965)

### 💡 Background and solution

Add Bin & KDE & Contour transform for VGrammar.

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### 🚀 Summary

copilot:summary

### 🔍 Walkthrough

copilot:walkthrough

### CODE REVIEW

File: packages/vgrammar/src/transforms/data/bin.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import type { BinTransformOption } from '../../types';
+
+export const transform = (options: BinTransformOption, upstreamData: any[]) => {
+  if (!upstreamData || upstreamData.length === 0) {
+    return upstreamData;
+  }
+
+  const { extent, bins, step } = options;
+  const range = extent[1] - extent[0];
+  const binStep = isValid(step) ? step : range / (bins ?? 10);
```
Comment (xile611) [inline packages/vgrammar/src/transforms/data/bin.ts:11]: `bins` maybe zero ?

File: packages/vgrammar/src/types/transform.ts [hunk 2]
```diff
@@ -163,6 +190,14 @@ export interface LttbSampleTransformOptions {
   groupBy?: string;
 }
 
+export interface BinTransformSpec extends BinTransformOption {
+  type: 'bind';
```
Comment (xile611) [inline packages/vgrammar/src/types/transform.ts:194]: the value of `type` is `bin`

File: packages/vgrammar/src/transforms/data/bin.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+  upstreamData.forEach(datum => {
```
Comment (xile611) [inline packages/vgrammar/src/transforms/data/bin.ts:13]: it's better to return a new array by default not modify the original data?
many other transforms have the same problem

File: packages/vgrammar/src/transforms/data/kde.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+};
+
+// A rule-of-thumb bandwidth estimator, referring to: https://en.wikipedia.org/wiki/Kernel_density_estimation#A_rule-of-thumb_bandwidth_estimator
+const ruleOfThumbBandwidth = (data: IPointLike[], dimension: number = 1) => {
+  const n = data.length;
+  const sum = data.reduce((sum, datum) => sum + datum.x, 0);
+  const mean = sum / n;
+  const sd = Math.sqrt(data.reduce((v, datum) => v + (datum.x - mean) ** 2, 0) / n);
+  const sortedData = data.sort((a, b) => a.x - b.x);
+  const q1 = (sortedData[Math.floor(n / 4)].x + sortedData[Math.ceil(n / 4)].x) / 2;
```
Comment (xile611) [inline packages/vgrammar/src/transforms/data/kde.ts:25]: vutils has a function `quantileSorted`, may be you can use `quantileSorted(sortedData, 0.25)` to replace this line

**Review by xile611 — CHANGES_REQUESTED** [review]

File: packages/vgrammar/src/transforms/data/bin.ts (hunk 1 above)
Comment (purpose233) [reply packages/vgrammar/src/transforms/data/bin.ts]: Fixed.

File: packages/vgrammar/src/transforms/data/bin.ts (hunk 3 above)
Comment (purpose233) [reply packages/vgrammar/src/transforms/data/bin.ts]: Fixed.

File: packages/vgrammar/src/transforms/data/kde.ts (hunk 4 above)
Comment (purpose233) [reply packages/vgrammar/src/transforms/data/kde.ts]: Fixed.

**Review by xile611 — APPROVED** [review]

**Pull request merged**